HeterogeneousCore/AlpakaCore/interface/ScopedContext.h

```diff
@@ -0,0 +1,300 @@
+#ifndef HeterogeneousCore_AlpakaCore_interface_ScopedContext_h
+#define HeterogeneousCore_AlpakaCore_interface_ScopedContext_h
+
+#include <memory>
+#include <stdexcept>
+#include <utility>
+
+#include "DataFormats/Portable/interface/Product.h"
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/ContextState.h"
+#include "HeterogeneousCore/AlpakaCore/interface/EventCache.h"
+#include "HeterogeneousCore/AlpakaCore/interface/QueueCache.h"
+#include "HeterogeneousCore/AlpakaCore/interface/ScopedContextFwd.h"
+#include "HeterogeneousCore/AlpakaCore/interface/chooseDevice.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/HostOnlyTask.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/traits.h"
+
+namespace cms::alpakatest {
+  class TestScopedContext;
+}
+
+namespace cms::alpakatools {
+
+  namespace impl {
+    // This class is intended to be derived by other ScopedContext*, not for general use
+    template <typename TQueue, typename>
+    class ScopedContextBase {
+    public:
+      using Queue = TQueue;
+      using Device = alpaka::Dev<Queue>;
+      using Platform = alpaka::Pltf<Device>;
+
+      Device device() const { return alpaka::getDev(*queue_); }
+
+      Queue& queue() const { return *queue_; }
```

Comment:
`Queue` is a `shared_ptr` to the actual object, so returning it by value or by reference has the same behaviour.
In fact, I'd prefer returning it by value -- but some alpaka functions take it by non-const reference, which cannot bind to a temporary, so that does not work.

Removing the `const` works, though, so fine by me to drop it.